prover/src/trace/trace_lde.rs

```diff
@@ -69,32 +69,30 @@ impl<E: FieldElement> TraceLde<E> {
     }
 
     /// Reads current and next rows from the main trace segment into the specified frame.
-    pub fn read_main_trace_frame_into(
+    pub fn read_main_trace_frame_into<F: EvaluationFrame<E::BaseField>>(
         &self,
         lde_step: usize,
-        frame: &mut EvaluationFrame<E::BaseField>,
+        frame: &mut F,
     ) {
         // at the end of the trace, next state wraps around and we read the first step again
-        let next_lde_step = (lde_step + self.blowup()) % self.trace_len();
+        // TODO: handle next step
+        let _next_lde_step = (lde_step + self.blowup()) % self.trace_len();
 
         // copy main trace segment values into the frame
-        self.main_segment_lde
-            .read_row_into(lde_step, frame.current_mut());
-        self.main_segment_lde
-            .read_row_into(next_lde_step, frame.next_mut());
+        frame.read_segment_into(lde_step, &self.main_segment_lde.columns);
```

Comment:
I think it should be the responsibility of the `EvaluationFrame` to read in the segment, especially if only certain row/column combinations are used. Is there a better way to expose the values in the main segment (and aux segments)?

Reply:
I agree that it should be the responsibility of the frame to read data from the trace. But this method should be generic enough to accommodate requirements of the prover (when we read frames from the original trace and from trace LDE) and the verifier (when we read frames from the proof).

Maybe taking an iterator over slices (where each slice is a column) could work. For example, something like:
```Rust
fn read_from<'a, I: Iterator<Item = &'a [E]>>(&mut self, columns: I, step: usize)
```  

Though, I think the above will have some issues with lifetimes.

Reply:
I agree that it should be the responsibility of the frame to read data from the trace. But this method should be generic enough to accommodate requirements of the prover (when we read frames from the original trace and from trace LDE) and the verifier (when we read frames from the proof).

Maybe taking an iterator over slices (where each slice is a column) could work. For example, something like:
```Rust
fn read_from<'a, I: Iterator<Item = &'a [E]>>(&mut self, columns: I, step: usize)
```  

Though, I think the above will have some issues with lifetimes.